File: wakakusa/lib/wakakusa/runtime.rb
# What the door holds between builds, and the calls the generated
# element methods make.
#
# There is one registry per kind of handler rather than one for all of
# them: a list of blocks that are all called with the same sort of value
# is a list a compiler can type, and a mixed one is not. The engine
# names the kind when it hands an event over, so the right registry is
# always the one that answers.
#
# A build starts the registries over, so a handler number means
# something only inside the build that handed it out. The engine holds
# no Ruby object at all: it has numbers, and asks for the rest.

# One registry, holding work to do rather than handlers to hand a value
# to. Every entry is a block of no arguments: the one an app wrote,
# wrapped in one that fetches what the event carried and passes it on.
#
# The reason is the compiler's. A list of blocks reached by index is a
# polymorphic list, and a call into one can only carry numbers — a
# string handed that way is read as an address, silently. Nothing but
# the wrapper ever crosses that call, and it takes no arguments at all.
$wakakusa_handlers = []
$wakakusa_rows = []
# The children being collected, when an element is being written as a
# block. Empty means the app is writing its children as arguments, and
# then nothing below does anything at all.
$wakakusa_frames = []
$wakakusa_row_index = 0
$wakakusa_app = nil
# Set once the window is up. A reload reads the app's file again, and
# the `run` at the bottom of it must not start a second one or throw
# away the object whose state the person has been building up.
$wakakusa_running = false
# Work the app started, and what it answered. The lock is the one place
# the window's thread and a worker meet.
$wakakusa_task_dones = []
$wakakusa_task_results = []
$wakakusa_task_lock = Mutex.new
$wakakusa_task_current = nil
# Timers are declared before the app runs and live for as long as it
# does, so they keep a list of their own that a build never clears.
$wakakusa_timers = []

# The text an event carried, built from the characters the engine counts
# out. It does not cross as a string for the same reason.
def wakakusa_event_text
  points = []
  n = PixieC.pixie_event_text_length
  i = 0
  while i < n
    points.push(PixieC.pixie_event_text_char(i))
    i += 1
  end
  points.pack("U*")
end

# The engine hands back the number an element carried; the work that was
# registered under it runs.
def wakakusa_on_event(id, _kind)
  $wakakusa_handlers[id].call
end

# Run `work` off the window's thread. When it is done the app's
# `on_done` method is called ON the window's thread, with what the work
# answered. Nothing inside the work may touch the app's state or the
# screen — the handler is where that belongs, and it is the reason the
# answer comes back this way rather than being written from the worker.
def task(&work)
  return 0 if work.nil?

  id = PixieC.pixie_task
  Thread.new do
    answer = work.call
    $wakakusa_task_lock.synchronize { $wakakusa_task_results[id] = answer }
    PixieC.pixie_task_done(id)
  end
  id
end

# What to do when that work is finished. The block runs on the window's
# thread, and `task_answer` inside it is what the work answered.
def on_done(job, &blk)
  return if blk.nil?

  $wakakusa_task_dones[job] = blk
end

# The engine is waiting for work and is handing us a turn. Without it
# nothing else in this program would ever run: the engine is on the
# stack from `run` until the window closes, and a thread scheduled by
# Ruby's own runtime gets no turn while that is true.
def wakakusa_pump
  Thread.pass
end

# The engine says a piece of work is finished, on the window's thread.
# The block is called with nothing and asks for the answer itself: what
# the work answered has no type known in advance, and a call whose
# argument cannot be typed is a call a compiled run cannot make.
def wakakusa_task_done(id)
  blk = $wakakusa_task_dones[id]
  return if blk.nil?

  $wakakusa_task_current = id.to_i
  blk.call
end

# What the work answered. Read it from the handler `on_done:` names.
def task_answer
  $wakakusa_task_lock.synchronize { $wakakusa_task_results[$wakakusa_task_current] }
end

# A timer came due.
def wakakusa_tick(id)
  $wakakusa_timers[id].call
end

# Ask to be told every `seconds`. Declared before `run`; both runs tick
# off the same clock, a frame in a window and an `advance:` in a script.
def every(seconds, &blk)
  return if blk.nil?

  $wakakusa_timers.push(blk)
  PixieC.pixie_every(seconds, $wakakusa_timers.length - 1)
end

# The engine asks for one row of a list that builds its rows on demand.
# The row number is left where the wrapper can read it rather than
# handed to the call: a list of blocks reached by index can only be
# called with nothing at all.
def wakakusa_row_build(handler, index)
  $wakakusa_row_index = index.to_i
  $wakakusa_rows[handler].call.to_i
end

# The engine asks for the tree.
def wakakusa_build
  $wakakusa_handlers = []
  $wakakusa_rows = []
  $wakakusa_frames = []
  $wakakusa_app.view
end

def wakakusa_work(el, key, work)
  $wakakusa_handlers.push(work)
  PixieC.pixie_on(el, key, $wakakusa_handlers.length - 1)
end

# One registration per kind of thing a handler is told, each wrapping
# the app's block in the fetch it needs.
def wakakusa_on_none(el, key, &blk)
  return if blk.nil?

  wakakusa_work(el, key, proc { blk.call })
end

def wakakusa_on_text(el, key, &blk)
  return if blk.nil?

  wakakusa_work(el, key, proc { blk.call(wakakusa_event_text) })
end

def wakakusa_on_bool(el, key, &blk)
  return if blk.nil?

  wakakusa_work(el, key, proc { blk.call(PixieC.pixie_event_int != 0) })
end

def wakakusa_on_int(el, key, &blk)
  return if blk.nil?

  wakakusa_work(el, key, proc { blk.call(PixieC.pixie_event_int) })
end

def wakakusa_on_float(el, key, &blk)
  return if blk.nil?

  wakakusa_work(el, key, proc { blk.call(PixieC.pixie_event_num) })
end

# The other way to name a handler: a symbol, which is the name of one of
# the app's own methods. A block cannot be handed through a keyword —
# a compiled run receives the address rather than the block, silently —
# so a second handler on the same element is written this way instead.
def wakakusa_send_none(el, key, name)
  wakakusa_work(el, key, proc { $wakakusa_app.send(name) })
end

def wakakusa_send_text(el, key, name)
  wakakusa_work(el, key, proc { $wakakusa_app.send(name, wakakusa_event_text) })
end

def wakakusa_send_bool(el, key, name)
  wakakusa_work(el, key, proc { $wakakusa_app.send(name, PixieC.pixie_event_int != 0) })
end

def wakakusa_send_int(el, key, name)
  wakakusa_work(el, key, proc { $wakakusa_app.send(name, PixieC.pixie_event_int) })
end

def wakakusa_send_float(el, key, name)
  wakakusa_work(el, key, proc { $wakakusa_app.send(name, PixieC.pixie_event_num) })
end

def wakakusa_rows(el, key, &blk)
  return if blk.nil?

  $wakakusa_rows.push(proc { blk.call($wakakusa_row_index) })
  PixieC.pixie_rows(el, key, $wakakusa_rows.length - 1)
end

# An element is finished. If a container is collecting, it joins that
# container's children; either way its handle is the answer.
def wakakusa_done(el)
  h = PixieC.pixie_end(el)
  frame = $wakakusa_frames.last
  frame.push(h) unless frame.nil?
  h
end

# The children of a container: the ones handed to it as arguments, then
# the ones its block wrote. An element handed over as an argument was
# already collected by whatever container is open, so it is taken back
# out — it belongs to this one now.
def wakakusa_collect(kids, &blk)
  frame = $wakakusa_frames.last
  unless frame.nil?
    kids.each { |h| frame.delete(h) }
  end
  return kids if blk.nil?

  $wakakusa_frames.push([])
  blk.call
  kids + $wakakusa_frames.pop
end

def wakakusa_children(el, kids)
  return if kids.empty?

  PixieC.pixie_children(el, wakakusa_ids(kids), kids.length)
end

def wakakusa_riders(el, riders, owns_label)
  riders.each do |name, v|
    if owns_label && name == :a11y_label
      raise ArgumentError, "this element's own `label` is already the name " \
                           "a screen reader reads; there is no second one to give"
    end
    wakakusa_rider(el, name, v)
  end
end

# Open the window and hand it the app: an object with a `view` method
# that answers one element. The app's state lives in its instance
# variables, and a handler is a block that closes over it.
#
# Under PIXIE_SCRIPT there is no window: the engine builds the tree,
# prints it, replays the script and returns.
def run(app, title: "wakakusa", width: 0.0, height: 0.0, padding: -1.0)
  return if $wakakusa_running

  $wakakusa_running = true
  $wakakusa_app = app
  wakakusa_start(title, width, height, padding)
end

File: gomamochi/internal/door/runtime.go
package door

import (
	"fmt"
	"os"
	"reflect"
	"runtime"
	"sync"
	"unsafe"

	"github.com/ebitengine/purego"
)

// What the door holds between builds. There is one registry of work to
// do, filled as the tree is written and started over at every build, so
// a handler number means something only inside the build that handed
// it out. Every entry is a closure of no arguments: the one the app
// wrote, wrapped in one that fetches what the event carried and passes
// it on. A Go closure knows its own type, so one registry serves every
// kind of handler.
var (
	handlers []func()
	rows     []func(int64) int64
	// Declared before the app runs, and living as long as it does.
	timers   []func()
	bindings []func()
	// Work the app started, and what it answered. The lock is the one
	// place a worker and the window's thread meet.
	taskMu  sync.Mutex
	answers = map[int64]any{}
	dones   = map[int64]func(any){}

	// The app the window is showing, how to build its tree, and
	// whether the window is already up: a reload re-reads the whole
	// file, and the second Run it reaches must not open a second
	// window or throw away the values a person has been building up.
	app     any
	build   func() int64
	running bool
	// The next Run is a re-read of the app's file: its app is the new
	// shape, and the old one's values are carried into it. While the
	// re-read declares its timers and bindings again, they take the
	// places of the ones already installed, in order.
	reloading          bool
	reTimer, reBinding int

	watchPath string
	reloadFn  func() bool
)

// Config is what Run is handed: the window, the app, and how to build
// its tree.
type Config struct {
	Title                  string
	Width, Height, Padding float64
	App                    any
	Build                  func() int64
}

// The callbacks the engine reaches the app through. Made once; each is
// a C function pointer for the life of the process.
var (
	cbBuild = purego.NewCallback(func() int64 {
		defer guard()
		handlers = handlers[:0]
		rows = rows[:0]
		return build()
	})
	cbEvent = purego.NewCallback(func(id, kind int64) {
		defer guard()
		if int(id) < len(handlers) {
			handlers[id]()
		}
	})
	cbRow = purego.NewCallback(func(h, i int64) int64 {
		defer guard()
		if int(h) < len(rows) {
			return rows[h](i)
		}
		return 0
	})
	cbTimer = purego.NewCallback(func(id int64) {
		defer guard()
		if int(id) < len(timers) {
			timers[id]()
		}
	})
	// The engine says a piece of work is finished, on the window's
	// thread; the app's closure is called with what the work answered.
	cbTask = purego.NewCallback(func(id int64) {
		defer guard()
		taskMu.Lock()
		done, answer := dones[id], answers[id]
		delete(dones, id)
		delete(answers, id)
		taskMu.Unlock()
		if done != nil {
			done(answer)
		}
	})
	// The engine is waiting for work and hands the app a turn. Go's
	// scheduler runs the app's goroutines on threads of its own, so
	// the turn is a courtesy here rather than a necessity.
	cbPump    = purego.NewCallback(func() { runtime.Gosched() })
	cbBinding = purego.NewCallback(func(id int64) {
		defer guard()
		if int(id) < len(bindings) {
			bindings[id]()
		}
	})
	// The app's file changed. The host reads it again; if that took,
	// the engine rebuilds the view.
	cbReload = purego.NewCallback(func() int32 {
		defer guard()
		if reloadFn == nil {
			return 0
		}
		reloading, reTimer, reBinding = true, 0, 0
		ok := reloadFn()
		reloading = false
		if ok {
			return 1
		}
		return 0
	})
)

// A panic inside a callback cannot unwind through the engine's frames,
// so it is said here and the process stops where it stands — which is
// what a refusal does on the other side of the face too.
func guard() {
	if e := recover(); e != nil {
		fmt.Fprintf(os.Stderr, "gomamochi: %v\n", e)
		os.Exit(1)
	}
}

// Run opens the window and hands it the app. Under PIXIE_SCRIPT there
// is no window: the engine builds the tree, prints it, replays the
// script and returns.
func Run(c Config) {
	if running {
		if reloading {
			carry(c.App, app)
			app, build = c.App, c.Build
		}
		return
	}
	ensure()
	running = true
	app, build = c.App, c.Build
	pixieSetEventHandler(cbEvent)
	pixieSetRowBuilder(cbRow)
	pixieSetTimerHandler(cbTimer)
	pixieSetTaskHandler(cbTask)
	pixieSetPumpHandler(cbPump)
	pixieSetBindingHandler(cbBinding)
	// Only an interpreted run watches: a compiled app is what it is.
	if watchPath != "" {
		pixieWatch(watchPath, cbReload)
	}
	pixieRun(c.Title, c.Width, c.Height, c.Padding, cbBuild)
}

// Watch asks Run to have the engine watch the app's file, and to call
// `reload` on the window's thread when it changes. `reload` reads the
// file again and answers whether that took; the Run it reaches then
// carries the app's values into the new shape.
func Watch(path string, reload func() bool) {
	watchPath, reloadFn = path, reload
}

// Handler files a closure under the next number and tells the engine
// that number for this key of this element.
func Handler(el int64, key int32, f func()) {
	handlers = append(handlers, f)
	pixieOn(el, key, int64(len(handlers)-1))
}

// RowBuilder is the same for a list that builds its rows on demand:
// the engine asks for row i and is answered with that row's handle.
func RowBuilder(el int64, key int32, f func(int64) int64) {
	rows = append(rows, f)
	pixieRows(el, key, int64(len(rows)-1))
}

// Every asks to be told every `seconds`. Declared before Run; both runs
// tick off the same clock, a frame in a window and an `advance:` in a
// script.
func Every(seconds float64, f func()) {
	if running {
		if reloading && reTimer < len(timers) {
			timers[reTimer] = f
		}
		reTimer++
		return
	}
	ensure()
	timers = append(timers, f)
	pixieEvery(seconds, int64(len(timers)-1))
}

// A shortcut, a menu item, a key or a dropped file: declared before the
// app runs, outliving every build, numbered in a list of their own.
func bind(f func()) (int64, bool) {
	if running {
		if reloading && reBinding < len(bindings) {
			bindings[reBinding] = f
		}
		reBinding++
		return 0, false
	}
	ensure()
	bindings = append(bindings, f)
	return int64(len(bindings) - 1), true
}

func Shortcut(chord string, f func()) {
	if id, fresh := bind(f); fresh {
		pixieShortcut(chord, id)
	}
}

func MenuItem(menu, item string, f func()) {
	if id, fresh := bind(f); fresh {
		pixieMenuItem(menu, item, id)
	}
}

func OnKey(f func()) {
	if id, fresh := bind(f); fresh {
		pixieOnKey(id)
	}
}

func OnFileDrop(f func()) {
	if id, fresh := bind(f); fresh {
		pixieOnFileDrop(id)
	}
}

// Task runs `work` on a goroutine of its own. When it is done, `done`
// is called on the window's thread with what the work answered.
// Nothing inside the work may touch the app or the screen: the answer
// comes back this way so that the handler is where the writing
// happens.
func Task(work func() any, done func(any)) {
	ensure()
	id := pixieTask()
	taskMu.Lock()
	dones[id] = done
	taskMu.Unlock()
	go func() {
		// A failing library call panics; off the window's thread the
		// same guard says it, or the runtime would print a trace.
		defer guard()
		v := work()
		taskMu.Lock()
		answers[id] = v
		taskMu.Unlock()
		pixieTaskDone(id)
	}()
}

// carry copies every field of the app the window had into the app the
// re-read file made, where the new one has a field of the same name and
// type; a field that changed its type starts over. The apps arrive as
// whatever the door was handed: a pointer to a struct, or an
// interpreter's wrapper around one.
func carry(newApp, oldApp any) {
	n := unwrap(reflect.ValueOf(newApp))
	o := unwrap(reflect.ValueOf(oldApp))
	if n.Kind() != reflect.Struct || o.Kind() != reflect.Struct {
		return
	}
	for i := 0; i < n.NumField(); i++ {
		f := n.Type().Field(i)
		of := o.FieldByName(f.Name)
		if !of.IsValid() || of.Type() != f.Type || !n.Field(i).CanAddr() || !of.CanAddr() {
			continue
		}
		dst := reflect.NewAt(f.Type, unsafe.Pointer(n.Field(i).UnsafeAddr())).Elem()
		src := reflect.NewAt(f.Type, unsafe.Pointer(of.UnsafeAddr())).Elem()
		dst.Set(src)
	}
}

func unwrap(v reflect.Value) reflect.Value {
	for {
		switch v.Kind() {
		case reflect.Interface, reflect.Pointer:
			if v.IsNil() {
				return v
			}
			v = v.Elem()
			continue
		case reflect.Struct:
			if v.NumField() >= 1 && v.Type().Field(0).Name == "IValue" {
				v = v.Field(0)
				continue
			}
		}
		return v
	}
}
